### PR TITLE
[Github #1323] get records pagination

### LINKED ISSUE #1323: Optimize getRecords pagination
Offset pagination has shown its performance limit. We need keyset pagination with proper indexes. 

What instrumentation & tracing to measure pagination bottleneck in the future?

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
Adds in a paginated response for the `listRecords` call. This added HTTP call and SDK method uses cursor pagination and sends a `next_cursor` property with a string or null value in the response. 

```
{
    records:
        [
            {
                id: 123,
                ..., // Fields as specified in the model you queried
                _nango_metadata: {
                    deleted_at: null,
                    last_action: 'ADDED',
                    first_seen_at: '2023-09-18T15:20:35.941305+00:00',
                    last_modified_at: '2023-09-18T15:20:35.941305+00:00'
                }
            },
            ...
        ],
    next_cursor: "Y3JlYXRlZF9hdF4yMDIzLTExLTE3VDExOjQ3OjE0LjQ0NyswMjowMHxpZF4xYTE2MTYwMS0yMzk5LTQ4MzYtYWFiMi1mNjk1ZWI2YTZhYzI"
}
```

## Issue ticket number and link
Resolves #1323 

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [x] I added tests 
- [x] I added observability
- [ ] I added analytics, otherwise the reason is: n/a

### CODE REVIEW

File: packages/node-client/lib/index.ts [hunk 1]
```diff
@@ -355,7 +365,40 @@ export class Nango {
...
+                ?.split('?')[1]
+                ?.split('&')
+                ?.find((param) => param.includes('cursor'));
+
+            const cursorOnly = cursor?.split('=')[1];
+
+            return cursorOnly || '';
+        };
+
+        return { records: response.data, next_cursor: parseLinkHeader(response.headers['link']) };
```
Comment (TBonnin) [inline packages/node-client/lib/index.ts]: what about also passing `next_cursor` as a custom header (ex: `x-nango-next` in addition to `links` so we don't have do all the parsing

**Review by TBonnin — APPROVED** [review]

File: docs-v2/sdks/node.mdx [hunk 2]
```diff
@@ -338,23 +336,32 @@ console.log(records);
...
+                _nango_metadata: {
+                    deleted_at: null,
+                    last_action: 'ADDED',
+                    first_seen_at: '2023-09-18T15:20:35.941305+00:00',
+                    last_modified_at: '2023-09-18T15:20:35.941305+00:00'
+                }
+            },
+            ...
+        ],
+    next_cursor: "Y3JlYXRlZF9hdF4yMDIzLTExLTE3VDExOjQ3OjE0LjQ0NyswMjowMHxpZF4xYTE2MTYwMS0yMzk5LTQ4MzYtYWFiMi1mNjk1ZWI2YTZhYzI"
```
Comment (bastienbeurier) [inline docs-v2/sdks/node.mdx:354]: Is the cursor still going to be this long??

Comment (khaliqgant) [reply docs-v2/sdks/node.mdx:354]: Yeah, the content is fairly long. Timestamp + uuid

for example:
```
2023-11-17T11:47:14.447+02:00||012d21e7-727a-446c-8a06-96847fa31f86
```

File: docs-v2/guides/integrate.mdx [hunk 3]
```diff
@@ -111,4 +111,4 @@ Dive deeper with the following guides:
 
 ## Questions, problems, feedback?
 
-We're here to help! Please reach out on the [Slack community](https://nango.dev/slack), we are very active there.
\ No newline at end of file
+We're here to help! Please reach out on the [Slack community](https://nango.dev/slack), we are very active there.We're here to help! Please reach out on the [Slack community](https://nango.dev/slack), we are very active there.
```
Comment (TBonnin) [inline docs-v2/guides/integrate.mdx]:
```suggestion
We're here to help! Please reach out on the [Slack community](https://nango.dev/slack), we are very active there.
```

File: packages/shared/lib/services/sync/data/records.service.ts [hunk 4]
```diff
@@ -290,6 +326,168 @@ export async function getDataRecords(
...
+        }
+
+        const customerResult = result.map((item) => item.record);
+
+        if (customerResult.length > (limit || 100)) {
+            customerResult.pop();
+            rawResult.pop();
+
+            const cursorRawElement = rawResult[rawResult.length - 1] as SyncDataRecord;
+            const cursorElement = customerResult[customerResult.length - 1] as unknown as CustomerFacingDataRecord;
```
Comment (TBonnin) [inline packages/shared/lib/services/sync/data/records.service.ts:476]: do I understand correctly the cursor is a pointer to the last record that is gonna be sent back?

**Review by TBonnin — APPROVED** [review]

File: packages/shared/lib/services/sync/data/records.service.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/shared/lib/services/sync/data/records.service.ts:476]: That is correct, and the next page query looks for records  > the provided cursor

File: docs-v2/guides/integrate.mdx (hunk 3 above)
Comment (khaliqgant) [reply docs-v2/guides/integrate.mdx]: Thanks, updated with https://github.com/NangoHQ/nango/pull/1350/commits/ddc130b489222219ee8464ce10d1a0cf88019f46

File: packages/shared/lib/services/sync/data/records.service.ts (hunk 4 above)
Comment (TBonnin) [reply packages/shared/lib/services/sync/data/records.service.ts:476]: I am nitpicking so you do as you prefer. `next` could actually be the first record following the last one of a page which would be more in phase with the name `next`. The condition would then be `>= provided cursor`.
Not big deal though if you prefer to keep it this way

Comment (khaliqgant) [reply packages/shared/lib/services/sync/data/records.service.ts:476]: Yeah, I see your point. Think I'll keep it like this for now.

**Pull request merged**